### DISCUSSION TITLE
MCP Protocol

### DISCUSSION THREAD

**Discussion by MrZLeo** [yetone/avante.nvim#908] [Ideas]
https://modelcontextprotocol.io/quickstart

Anyone hear about MCP protocol proposed by [AnthropicAI](https://x.com/AnthropicAI)? Can we provide a server like this to enhance experience?

**Comment by ndom91** [comment]
Came here searching for just this post. Yeah I think this would make a great match for Avante. Will keep my eyes on this closely!

For some additional context, here's their intro blog post about it: https://www.anthropic.com/news/model-context-protocol:

> Today, we're open-sourcing the [Model Context Protocol](https://modelcontextprotocol.io/) (MCP), a new standard for connecting AI assistants to the systems where data lives, including content repositories, business tools, and development environments. Its aim is to help frontier models produce better, more relevant responses...

**Comment by dogonthehorizon** [comment]
So if I understand correctly basically what this does is expose a set of tools on an api endpoint that other applications can integrate with and include in prompts sent to the LLM? If that's true then what I understand the architecture would be to have avante initialize/discover mcp servers and then make sure those tools are included when it sends avante ask and/or completion requests to the LLM.

I imagine that would require an ecosystem of mcp servers to be built out to handle a variety of tasks (local code search, documentation search, etc).

From a installation/maintenance standpoint also not sure what the best path would be. The reference implementations are python/typescript but the protocol is open. Should the decision be made to support it, it would probably make more sense to do the mcp brokering in lua and have a way for mason to install mcp servers similar to what it does for lsps/daps/etc.

**Reply by MrZLeo** [reply]
I think as an AI agent, avante.nvim should provide a framework similar to Claude.app to allow users to attach different tools they want. We don't need to care about what the actual tools are.

One obstacle is that we don't have a Lua SDK currently. However, we can directly implement a Lua library and transfer to a de facto SDK once it becomes available (or create our own!), since it just involves things like JSON and HTTP.

One reference is Zed, which has MCP support right now. Zed is built with Rust, which means they already have a Rust version of the SDK (which may be released in the future?). Can we integrate the Rust SDK directly?​​​​​​​​​​​​​​​​

**Reply by dogonthehorizon** [reply]
I've been thinking about this a lot the last three weeks @MrZLeo. One way to shortcircuit support for MCP would be to [wrap the python-sdk](https://github.com/modelcontextprotocol/python-sdk) in [neovim's rplugin system](https://pynvim.readthedocs.io/en/latest/usage/remote-plugins.html) so we can get to an avante implementation quicker.

I think this makes sense as a standalone plugin and then build an interface in avante that knows how to access tools/resources exposed by MCP servers.

Is that in line with how you've been thinking about this too?

**Comment by aarnphm** [comment]
I think MCP can be an extensions of tool use. At the end of the day they are built on top of how these models return structured outputs.
